src/class-model.ts

```diff
@@ -0,0 +1,355 @@
+  const metadatas = tags ? getExplicitMetadataFromTags(tags) : getReflectionMetadata(klass);
+  const explicitKeys = new Set<string>(metadatas.map((metadata) => metadata.property));
+
+  for (const property of allPrototypeFunctionProperties(klass.prototype)) {
+    if (explicitKeys.has(property)) continue;
+    metadatas.push({
+      type: "view",
+      property,
+    });
+  }
+
+  for (const metadata of metadatas) {
+    switch (metadata.type) {
+      case "view": {
+        const descriptor = Object.getOwnPropertyDescriptor(klass.prototype, metadata.property);
+        if (!descriptor) {
+          throw new RegistrationError(`Property ${metadata.property} not found on ${klass} prototype, can't register view for class model`);
+        }
+        Object.defineProperty(mstViews, metadata.property, {
+          ...Object.getOwnPropertyDescriptor(klass.prototype, metadata.property),
+          enumerable: true,
+        });
+        break;
+      }
+      case "action": {
+        let target: any;
+        if (metadata.property in klass.prototype) {
+          target = klass.prototype;
+        } else {
+          // hackily instantiate the class to get at the instance level properties defined by the class body (those that aren't on the prototype)
+          target = new (klass as any)({}, undefined, undefined, true);
+        }
+        const descriptor = Object.getOwnPropertyDescriptor(target, metadata.property);
+
+        if (!descriptor) {
+          throw new RegistrationError(
+            `Property ${metadata.property} not found on ${klass} prototype or instance, can't register action for class model`
+          );
+        }
+
+        // add the action to the MST actions we'll add to the MST model type
+        Object.defineProperty(mstActions, metadata.property, {
+          ...descriptor,
+          enumerable: true,
+        });
+
+        // mark the action as not-runnable on the readonly class
+        Object.defineProperty(klass.prototype, metadata.property, {
+          ...descriptor,
+          enumerable: true,
+          value: defaultThrowAction(metadata.property),
+        });
+
+        break;
+      }
+      case "volatile": {
+        mstVolatiles[metadata.property] = metadata;
+      }
+    }
+  }
+
+  klass.volatiles = mstVolatiles;
+
+  // conform to the API that the other MQT types expect for creating instances
+  klass.instantiate = (snapshot, context) => new klass(snapshot, context.env, context);
+  (klass as any).is = (value: any) => value instanceof klass || klass.mstType.is(value);
+  klass.create = (snapshot, env) => klass.mstType.create(snapshot, env);
+  klass.createReadOnly = (snapshot, env) => new klass(snapshot, env) as any;
+
+  // create the MST type for not-readonly versions of this using the views and actions extracted from the class
+  klass.mstType = mstTypes
+    .model(klass.name, mstPropsFromQuickProps(klass.properties))
```

Comment:
It's not really clear to me how this name is initialized, it looks like it's coming from the object returned from `ClassModel` and not the class declaration.

Reply:
Should this be `klass.__proto__.name` instead?

Reply:
I still feel like this should be using `Reflect.getPrototypeOf(klass).constructor.name` - otherwise I think all class models will have `AnonymousModel` for the name.